state-synchronizer/src/peer_manager.rs

```diff
@@ -122,52 +126,80 @@ impl PeerManager {
-            match WeightedIndex::new(&weights) {
-                Ok(weighted_index) => {
-                    self.weighted_index = Some(weighted_index);
-                }
-                Err(e) => {
-                    error!(
-                        "[state sync] (pick_peer) failed to compute weighted index, {:?}",
-                        e
-                    );
-                }
-            }
-        }
+        // compute weighted index based on updated eligible peers
+        let mut eligible_peers = vec![];
+        let weights: Vec<_> = active_peers
+            .into_iter()
+            .map(|(peer, peer_info)| {
+                eligible_peers.push(peer.clone());
+                peer_info.score
+            })
+            .collect();
+        self.eligible_peers = eligible_peers;
+        self.weighted_index = WeightedIndex::new(&weights)
+            .or_else(|err| {
+                error!(
+                    "[state sync] (pick_peer) failed to compute weighted index, {:?}",
+                    err
+                );
+                Err(err)
+            })
+            .ok();
     }
 
     pub fn pick_peer(&self) -> Option<PeerNetworkId> {
-        let active_peers = self.get_active_upstream_peers();
-        debug!("[state sync] (pick_peer) state: {:?}", self.peers);
-
         if let Some(weighted_index) = &self.weighted_index {
             let mut rng = thread_rng();
-            if let Some(peer) = active_peers.get(weighted_index.sample(&mut rng)) {
-                return Some(peer.0.clone());
+            if let Some(peer) = self.eligible_peers.get(weighted_index.sample(&mut rng)) {
+                return Some(peer.clone());
             }
         }
         None
     }
 
     fn get_active_upstream_peers(&self) -> Vec<(&PeerNetworkId, &PeerInfo)> {
-        self.peers
-            .iter()
-            .filter(|&(peer, peer_info)| peer_info.is_alive && self.is_upstream_peer(peer))
-            .collect()
+        if self.upstream_config.networks.len() > 1 {
+            // failover mode is enabled only if there are multiple upstream networks
+            // in failover mode, we select the network of the highest preference (defined by UpstreamConfig)
+            // with at least one live peer
+            // We failover to the next network (in order of preference defined by UpstreamConfig) if
+            // there are no live peers in each network
+
+            // group active upstream peers by network
+            let active_peers_by_network = self
+                .peers
+                .iter()
+                .filter(|(_peer, peer_info)| peer_info.is_alive)
+                .map(|(peer, peer_info)| (peer.network_id(), (peer, peer_info)))
+                .into_group_map();
+
+            // find the first network with any live peers
+            self.upstream_config
+                .networks
+                .iter()
+                .find_map(|network| active_peers_by_network.get(network))
```

Comment:
it looks to me the ranking follows the order of the positions, do we want to align with `get_upstream_preference`?

e.g. if upstream_config.network is [fullnode 1, fullnode 2, validator], they return different result.

Reply:
it's part of the config req that the `networks` in the upstream config are already ordered by preference